src/realm/replication.cpp

```diff
@@ -62,36 +64,323 @@ Replication::version_type Replication::prepare_commit(version_type orig_version)
+                        t->get_class_name(), col_name, collection_type, target_table->get_class_name());
+        }
+        else {
+            logger->log(util::Logger::Level::debug, "On class '%1': Add property '%2' %3of %4", t->get_class_name(),
+                        col_name, collection_type, type);
+        }
+    }
+    select_table(t);                  // Throws
+    m_encoder.insert_column(col_key); // Throws
+}
+
+void Replication::erase_column(const Table* t, ColKey col_key)
+{
+    if (auto logger = get_logger()) {
+        logger->log(util::Logger::Level::debug, "On class '%1': Remove property '%2'", t->get_class_name(),
+                    t->get_column_name(col_key));
+    }
+    select_table(t);                 // Throws
+    m_encoder.erase_column(col_key); // Throws
+}
+
 void Replication::create_object(const Table* t, GlobalKey id)
 {
+    if (auto logger = get_logger()) {
+        logger->log(util::Logger::Level::debug, "Create object '%1'", t->get_class_name());
+    }
     select_table(t);                              // Throws
     m_encoder.create_object(id.get_local_key(0)); // Throws
 }
 
-void Replication::create_object_with_primary_key(const Table* t, ObjKey key, Mixed)
+void Replication::create_object_with_primary_key(const Table* t, ObjKey key, Mixed pk)
 {
+    if (auto logger = get_logger()) {
+        logger->log(util::Logger::Level::debug, "Create object '%1' with primary key %2", t->get_class_name(), pk);
+    }
     select_table(t);              // Throws
     m_encoder.create_object(key); // Throws
 }
 
+void Replication::remove_object(const Table* t, ObjKey key)
+{
+    if (auto logger = get_logger()) {
+        if (t->is_embedded()) {
+            logger->log(util::Logger::Level::debug, "Remove embedded object '%1'", t->get_class_name());
+        }
+        else if (t->get_primary_key_column()) {
+            logger->log(util::Logger::Level::debug, "Remove object '%1' with primary key %2", t->get_class_name(),
+                        t->get_primary_key(key));
+        }
+        else {
+            logger->log(util::Logger::Level::debug, "Remove object '%1'", t->get_class_name());
+        }
+    }
+    select_table(t);              // Throws
+    m_encoder.remove_object(key); // Throws
+}
+
+inline void Replication::select_obj(ObjKey key)
+{
+    if (key != m_selected_obj) {
+        if (auto logger = get_logger()) {
+            if (logger->would_log(util::Logger::Level::debug)) {
+                auto class_name = m_selected_table->get_class_name();
+                if (m_selected_table->get_primary_key_column()) {
+                    auto pk = m_selected_table->get_primary_key(key);
+                    logger->log(util::Logger::Level::debug, "Mutating object '%1' with primary key %2", class_name, pk);                   
+                }
+                else if (m_selected_table->is_embedded()) {
```

Comment:
is this the path objects in nested collections will take as well? Or do we need different handling?

Reply:
Yes - if a nested collection is modified, it is always the object owning the collection tree that is selected.